packages/mesos-client/src/rxjs-extensions/keepAlive.js

```diff
@@ -0,0 +1,111 @@
+import { Subscriber, Scheduler } from "rxjs";
+
+export default function keepAlive(
+  source,
+  backoffInterval = 5000,
+  maxRetries = -1,
+  scheduler = Scheduler.async
+) {
+  return source.lift(
+    new KeepAliveOperator(source, backoffInterval, maxRetries, scheduler)
+  );
+}
+
+class KeepAliveOperator {
+  constructor(source, backoffInterval, maxRetries, scheduler) {
+    this.source = source;
+    this.maxRetries = maxRetries;
+    this.backoffInterval = backoffInterval;
```

Comment:
The keepalive operartor shouldn't also apply the backoff. Every operator should only to one thing. 
 
It should be possible to add the backoff using `concat` as illustrated in the following diagram: 

![](https://user-images.githubusercontent.com/647035/30740916-2d67aec8-9f93-11e7-89ef-7800d21cf251.jpg)

Reply:
Then `keepAlive` is just a combination of `retry()` and `repeat()` :)
and, as you might remember, in the previous commit that I've kept on purpose, we've had the backoff extracted and it was rather confusing.

Reply:
Regarding the `concat` operator, isn't it that it should know all the observables upfront? We need to resubscribe to the stream dynamically.

Reply:
Yep, we don't even need  `retry` as we catch the connection errors.

```js 
  Rx.Observable.create((observer) => {
    console.log("subscribe")
    setTimeout(() => {
      if(Math.random() >= 0.5){
         console.log("complete")
         observer.complete()
      } else {
        console.log("error")
        observer.error()
      }
    }, 1000)
  })
  .catch((err) => {
      return Rx.Observable.empty();
  })
  .concat(
    Rx.Observable.create( (observer) => {
      console.log("backoff")
      setTimeout(function (){
        observer.complete();
      }, 1000)
    })
  )
  .repeat()
  .subscribe()
```

Reply:
@MatApple would like to get your view on both version.